lib/rucio/api/replica.py

```diff
@@ -102,18 +104,22 @@ def declare_bad_file_replicas(replicas, reason, issuer, vo='def', session=None):
             if rse_id is None:
                 rse = r["rse"]
                 rse_map[rse] = rse_id = get_rse_id(rse=rse, vo=vo, session=session)
+                rse_id_to_name[rse_id] = rse
             replicas_lst.append({
                 "scope": scope,
                 "rse_id": rse_id,
                 "name": r["name"]
             })
-    undeclared = replica.declare_bad_file_replicas(replicas_lst, reason=reason, issuer=issuer, status=BadFilesStatus.BAD, session=session)
+    undeclared = replica.declare_bad_file_replicas(replicas_lst, reason=reason, issuer=issuer, status=BadFilesStatus.BAD,
+                                                   force=force, session=session)
     out = {}
     for rse_id, ulist in undeclared.items():
         if ulist:
             rse_name = None
             if rse_id == 'unknown':
                 rse_name = 'unknown'
+            elif rse_id in rse_id_to_name:
+                rse_name = rse_id_to_name[rse_id]
```

Comment:
After a quick talk with @faluchet, we think that the `rse_id_to_name` could be redundant.

utils.py:
```python
def invert_dict(d):
  return {value: key for (key, value) in d.items()}
```

api/replica.py
```python
elif rse_id in rse_map.values():
  rse_name = invert_dict(rse_map)[rse_id]
```
not work? This way we don't store information redundant. The performance added is negligible small.

Reply:
I do not see invert_dict in utils...
Are you proposing to invert the dictionary every time the mapping from rse_id to name is needed ?
I think I will disagree with you.

Reply:
`inverted_dict` does not exist at this point, however it's a simple one-liner and adds a lot of value. I'd propose to create it.

`rse_map` is not changed in the loop, and the inverted dict is not needed before it. The inverted dictionary can be created before it, thus not, it does not have to be created every time. Just once.

```python
    rse_id_to_name = invert_dict(rse_map)
    for rse_id, ulist in undeclared.items():
        if ulist:
            rse_name = None
            if rse_id == 'unknown':
                rse_name = 'unknown'
            elif rse_id in rse_id_to_name:
                rse_name = rse_id_to_name[rse_id]
```

The part I don't get is this new line:
```diff
                 except exception.RSENotFound:
                     rse_name = str(rse_id)
                 except ValueError:
                     rse_name = str(rse_id)
+                rse_id_to_name[rse_id] = rse_name
             if rse_name:
                 out[rse_name] = out.get(rse_name, []) + ulist
```
`rse_id` is unique, it's the key of the dictionary. Setting it is unnecessary from my POV.

About the timing point: Don't optimize code for performance that's not clearly needed. However, optimize it for readability. Using two dictionaries instead of one is not readable at all from my POV.

Reply:
I thought originally you proposed to inverse the dictionary in the loop where it is used (not where it is created) and I do not think that was a good idea.

Now it seems you are proposing to invert it once, before the loop where it is needed ? Please clarify.

That new line is how the mapping from rse_id to rse name is filled. Or now it is my turn to be missing something ?

Reply:
And speaking of performance, I know that at least in the use case I am working on it is clearly needed.

Reply:
Yes, in the beginning I did propose to add it into the loop. However, I realized that it can be added before it.

> Now it seems you are proposing to invert it once, before the loop where it is needed ?

Yes.

> That new line is how the mapping from rse_id to rse name is filled.

I think this line (line 131 of the `api/replcia.py` **after** your change) is useless. You set the value for the `rse_id` key in the `rse_id_to_name` dictionary, but since the `rse_id` is unique (it is the key in the `undeclared` dictionary, the one which items we iterate over) it will never be read. (Each `rse_id` will get processed exactly _once_ in the for loop, thus writing it there makes no sense.)
We only read `rse_id_to_name` in the for loop, thus we can generate it before, from the `rse_map` variable.

Reply:
Thanks @joeldierkes  for bringing the conversation forward.

@ivmfnal  thanks for the edits so far, just two notes:

a) From a user perspective it is not clear what happens when a caller sets both `rse_id` and `rse`, possibly mismatching.
L103 reveals that `rse_id` takes always priority, but unless one inspects the code it is definitely not clear. Is this the intended behavior, or perhaps the call should fail? Would you mind clarify, at least with a comment on the function documentation?

b) Also, I find inconsistent how the `rse_id`s returned from `core.declare_bad_file_replicas` are checked against exceptions in L125, while the `rse` coming as parameter from the function caller are not in L105. Typically, it is desirable to early sanitize inputs, instead of their by-products. Actually, I am not even sure that the L125 try-except block would ever trigger. Please can you address that?


Other than that, it looks like accommodating both a list of strings (pfns) and a list of dictionaries (replicas metadata) as parameters of the same function made this part of the code a bit convoluted, but this is not the place to address it.

JM2C